### DISCUSSION TITLE
[A-L] (10 aprile 2024) Esercizio 3

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#39] [Esercitazioni A-L]
Rimuovere un arco da un grafo (diretto o non diretto) modifica le distanze tra alcune coppie di vertici e ne lascia invariate alcune.
Fornire un algortimo in pseudo-codice che dato un grafo $G = (V, E)$ diretto, un vertice $s \in V$ , un arco $(u, v) \in E$ e un vettore dei padri $P$, relativo ad una BFS effettuata su $G$ a partire da $s$, determina se la rimozione di $(u, v)$ da $G$ modifica le distanze da $s$. È possibile ottenere una soluzione con tempo di esecuzione $O(|V|)$?

**Comment by LeoCosta02** [comment]
```C
/*Fornire un algortimo in pseudo-codice che dato un grafo G = (V, E) diretto,
un vertice s ∈ V , un arco (u, v) ∈ E e un vettore dei padri P arent, relativo
ad una BFS effettuata su G a partire da s determina se la rimozione di (u, v)
da G modifica le distanze da s. E’ possibile ottenere una soluzione con tempo
di esecuzione O(|V |)?*/

Main(G,s,(u,v),P){
 int dist[n]=[-1...-1]
 global int check[n]=[0...0]
    if(P[v]!=u) // arco (u,v) non appartiene all'albero -> la distanza da s non cambia 
     return false
      if(P[v]==u AND v.degIN()==0)// se (u,v) è un arco dell'albero e v non ha archi entranti allora se elimino tale arco cambia la distanza da s 
       return true
   
        for (int i=0;i<P.size();i++){
         if(check[i]==0)
         Dist_ric(P,dist,i)
        }
         for each(x entrante in v ){ // per ogni nodo che ha un arco entrante del tipo (x,v) se la distanza di x è diversa da u allora cambia la distanza da s 
            if(dist[x] != dist[u])
             return true
         }
          return false 
}

Dist_ric(P,dist,x){
    if(P[x]==x){
        dist[x]=0
        check[x]=1
        return 0
    }
     if(dist[P[x]]>=0){//se hai gia cacolato la distanza del padre
      dist[x]=dist[P[x]]+1
      check[x]=1
      return dist[x]
     }
      dist[x]=Dist_ric(P,dist,P[x])+1
      check[x]=1
      return dist[x]
}
```

**Reply by LeoCosta02** [reply]
Verificare se ho considerato tutti i casi possibili per determinare se la distanza cambia

**Comment by daromacarlo** [comment]
Riaggancio il nodo slegato al grafo con l'arco entrante a v che abbia il nodo più vicino possibile ad s, trovo il vettore delle distanze sia per il grafo senza modica che per quello modificato, controllo infine se sono uguali, se si ritorno true, altrimenti false. L'algoritmo opera in O(n)
![immagine](https://github.com/user-attachments/assets/5cc1224a-159b-4838-a865-2cb63a4acde7)
![immagine](https://github.com/user-attachments/assets/9c8178f8-f250-4846-bfb3-92a44cd0cf9a)
